pkg/agent/multicast/vif_allocator.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicast
+
+import (
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+
+	multicastsyscall "antrea.io/antrea/third_party/unix"
+
+	"k8s.io/klog/v2"
+)
+
+// VIF stands for virtual interface for multicast in kernel multicast implementation
+// and is used for adding/deleting static multicast entries, please check
+// https://github.com/torvalds/linux/blob/master/include/uapi/linux/mroute.h#L80
+// VIFAllocator is used for saving vif <-> interface mapping.
+type VIFAllocator struct {
+	mutex  sync.RWMutex
+	vifMap map[string]uint16
+	sockFD int
+}
+
+func newVIFAllocator(sockFD int) *VIFAllocator {
+	return &VIFAllocator{
+		vifMap: make(map[string]uint16),
+		sockFD: sockFD,
+	}
+}
+
+func (v *VIFAllocator) getVif(name string) (uint16, error) {
+	v.mutex.RLock()
+	defer v.mutex.RUnlock()
+	vif, ok := v.vifMap[name]
+	if !ok {
+		return 0, fmt.Errorf("VIF for interface %s not found", name)
+	}
+	return vif, nil
+}
+
+func (v *VIFAllocator) allocateVif(name string) (uint16, error) {
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return 0, err
+	}
+	for _, iface := range ifaces {
+		if strings.Contains(iface.Flags.String(), MulticastFlag) && iface.Name == name {
+			vif, err := v.allocate(iface.Index)
+			if err != nil {
+				return 0, err
+			}
+			err = setVifToInterface(v.sockFD, vif, iface.Index)
+			if err != nil {
+				return 0, err
+			}
+			klog.V(2).InfoS("Successfully allocate VIF", "VIF", vif)
+			v.vifMap[name] = vif
+			return vif, nil
+		}
+	}
+	return 0, fmt.Errorf("error adding VIF for %s failed: cannot find this interface", name)
+}
+
+func (v *VIFAllocator) setVifIfNotExist(name string) (uint16, error) {
+	klog.V(2).InfoS("Get and set VIF", "VIF", name)
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+	vif, exist := v.vifMap[name]
+	if !exist {
+		newvif, err := v.allocateVif(name)
+		if err != nil {
+			return 0, err
+		}
+		return newvif, nil
+	}
+	return vif, nil
+}
+
+func (v *VIFAllocator) allocate(ifIdx int) (uint16, error) {
+	klog.V(2).InfoS("Allocating VIF", "ifIdx", ifIdx)
```

Comment:
The method doesn't really need ifIdx. Even it's used for logging, why index is used, instead of the more readable interface name?

Reply:
reminder

Reply:
This method has been removed. Increment vif by 1 in the for loop is sufficient.